internal/crypto/certgen_test.go

```diff
@@ -0,0 +1,133 @@
+package crypto
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/envoyproxy/gateway/api/config/v1alpha1"
+)
+
+func TestGenerateCerts(t *testing.T) {
+	type testcase struct {
+		envoyGateway            *v1alpha1.EnvoyGateway
+		certConfig              *Configuration
+		wantEnvoyGatewayDNSName string
+		wantEnvoyDNSName        string
+	}
+
+	run := func(t *testing.T, name string, tc testcase) {
+		t.Helper()
+
+		t.Run(name, func(t *testing.T) {
+			t.Helper()
+
+			got, err := GenerateCerts(tc.envoyGateway)
+			require.NoError(t, err)
+
+			roots := x509.NewCertPool()
+			ok := roots.AppendCertsFromPEM(got.CACertificate)
+			require.Truef(t, ok, "Failed to set up CA cert for testing, maybe it's an invalid PEM")
+
+			currentTime := time.Now()
+
+			err = verifyCert(got.EnvoyGatewayCertificate, roots, tc.wantEnvoyGatewayDNSName, currentTime)
+			assert.NoErrorf(t, err, "Validating %s failed", name)
+
+			err = verifyCert(got.EnvoyCertificate, roots, tc.wantEnvoyDNSName, currentTime)
+			assert.NoErrorf(t, err, "Validating %s failed", name)
+		})
+	}
+
+	run(t, "no configuration - use defaults", testcase{
+		certConfig:              &Configuration{},
+		wantEnvoyGatewayDNSName: "envoy-gateway",
+		wantEnvoyDNSName:        "envoy",
+	})
+}
+
+func TestGeneratedValidKubeCerts(t *testing.T) {
+	now := time.Now()
+	expiry := now.Add(24 * 365 * time.Hour)
+
+	caCert, caKey, err := newCA("envoy-gateway", expiry)
+	require.NoErrorf(t, err, "Failed to generate CA cert")
+
+	egCertReq := &certificateRequest{
+		caCertPEM:  caCert,
+		caKeyPEM:   caKey,
+		expiry:     expiry,
+		commonName: "envoy-gateway",
```

Comment:
Common names are deprecated in RFC 2818 (in 2000) in favor of SANs. Major softwares (Golang, Chrome etc) stopped supporting them, so we should just issue certs without common names.